framework/src/executioners/MultiSystemSolveObject.C

```diff
@@ -0,0 +1,60 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "MultiSystemSolveObject.h"
+#include "FEProblemBase.h"
+#include "LinearSystem.h"
+#include "NonlinearSystem.h"
+
+InputParameters
+MultiSystemSolveObject::validParams()
+{
+  InputParameters params = emptyInputParameters();
+  params.addParam<std::vector<SolverSystemName>>(
+      "system_names",
+      "Names of the solver systems (both linear and nonlinear) that will be solved");
+  params.addParamNamesToGroup("system_names", "Multiple solver system");
+  return params;
+}
+
+MultiSystemSolveObject::MultiSystemSolveObject(Executioner & ex) : SolveObject(ex)
+{
+  // Retrieve pointers to all the systems from the problem by default
+  const auto & nl_sys_names = _problem.getNonlinearSystemNames();
+  const auto & linear_sys_names = _problem.getLinearSystemNames();
+  if (!isParamValid("system_names"))
+  {
+    for (const auto & sys_name : nl_sys_names)
+      _systems.push_back(&_problem.getSolverSystem(_problem.nlSysNum(sys_name)));
+    for (const auto & sys_name : linear_sys_names)
+      _systems.push_back(&_problem.getSolverSystem(_problem.nlSysNum(sys_name)));
+    _num_nl_systems = nl_sys_names.size();
+  }
+  else
+  {
+    _num_nl_systems = 0;
+    // Retrieve pointers to all the user-specified systems in the order that the user specified them
+    for (const auto & sys_name : getParam<std::vector<SolverSystemName>>("system_names"))
+    {
+      if (std::find(nl_sys_names.begin(), nl_sys_names.end(), sys_name) != nl_sys_names.end())
+      {
+        _systems.push_back(&_problem.getSolverSystem(_problem.nlSysNum(sys_name)));
+        _num_nl_systems++;
+      }
+      else if (std::find(linear_sys_names.begin(), linear_sys_names.end(), sys_name) !=
+               linear_sys_names.end())
+        _systems.push_back(&_problem.getSolverSystem(_problem.nlSysNum(sys_name)));
+      else
+        paramError("system_names",
+                   "System '" + sys_name +
+                       "' was not found in the Problem. Did you forget to declare it in the "
+                       "[Problem] block?");
+    }
+  }
```

Comment:
```suggestion
  // Retrieve pointers to all the systems from the problem by default
  const auto & solver_sys_names = _problem.getSolverSystemNames();
  const auto & linear_sys_names = _problem.getLinearSystemNames();
  if (!isParamValid("system_names"))
  {
    Add accessor in FEProblemBase for the stored vector<SolverSystem> and just use those
  }
  else
  {
    _num_nl_systems = 0;
    // Retrieve pointers to all the user-specified systems in the order that the user specified them
    for (const auto & sys_name : getParam<std::vector<SolverSystemName>>("system_names"))
    {
      if (std::find(nl_sys_names.begin(), nl_sys_names.end(), sys_name) != nl_sys_names.end())
      {
        _systems.push_back(&_problem.getSolverSystem(_problem.solverSysNum(sys_name)));
        _num_nl_systems++;
      }
      else if (std::find(linear_sys_names.begin(), linear_sys_names.end(), sys_name) !=
               linear_sys_names.end())
        _systems.push_back(&_problem.getSolverSystem(_problem.solverSysNum(sys_name)));
      else
        paramError("system_names",
                   "System '" + sys_name +
                       "' was not found in the Problem. Did you forget to declare it in the "
                       "[Problem] block?");
    }
  }
```

Reply:
I am fine if you keep the current method to be honest I was just thinking about a loop over solver systems instead of splitting it in two. 

The more relevant parts of the change are these:
```
_systems.push_back(&_problem.getSolverSystem(_problem.solverSysNum(sys_name)));
```
because of the same numbering shift we discussed below.

Reply:
good catch.
I think we dont need nlSysNum anymore if the "nonlinear system index" and the "solver system index" match

Reply:
I ll need to add a test for multi-linear system to avoid these problems. I d rather do it in a later PR

Maybe a DiffusionLinearFV physics would be a good excuse to add that?